aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -178,4 +163,207 @@ Tensor mm_cuda(const Tensor& self, const Tensor& mat2) {
+              vec1.device(), " and ", vec2.device());
+  TORCH_CHECK(self_.dim() == 2,
+              "2D tensor expected, got ", self_.dim(), "D tensor for input");
+  TORCH_CHECK(self_.size(0) == vec1.size(0) && self_.size(1) == vec2.size(0),
+              "size mismatch",
+              ", input: ", self_.sizes(),
+              ", v1: ", vec1.sizes(),
+              ", v2: ", vec2.sizes());
+  AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, self_.scalar_type(), "addr_out_cuda", [&] {
+      addr_impl_cuda<scalar_t>(out, self_, vec1, vec2,
+                               alpha.to<scalar_t>(), beta.to<scalar_t>());
+  });
+  return out;
+}
+
+Tensor& addr__cuda(Tensor& self,
+                   const Tensor& vec1, const Tensor& vec2,
+                   Scalar beta, Scalar alpha) {
+  addr_out_cuda(self, self, vec1, vec2, beta, alpha);
+  return self;
+}
+
+Tensor addr_cuda(const Tensor& self,
+                  const Tensor& vec1, const Tensor& vec2,
+                  Scalar beta, Scalar alpha) {
+  Tensor out = at::empty({0}, self.options());
+  addr_out_cuda(out, self, vec1, vec2, beta, alpha);
+  return out;
+}
+
+Tensor& addbmm_out_cuda(Tensor& out, const Tensor& self,
+                        const Tensor& batch1, const Tensor& batch2,
+                        Scalar beta, Scalar alpha) {
+  TORCH_CHECK(batch1.dim() == 3 && batch2.dim() == 3,
+              "Batch tensors should be 3D, got dimensions ", batch1.dim(),
+              " and ", batch2.dim());
+
+  Tensor self_;
+  if (&out != &self) {
+    std::tie(self_) = expand_size(self, {batch1.size(1), batch2.size(2)}, "addbmm");
+  } else {
+    self_ = self;
+  }
+
+  TORCH_CHECK(out.device() == self_.device() &&
+              out.device() == batch1.device() &&
+              out.device() == batch2.device(),
+              "Expected all tensors to be on the same device. Found: ",
+              out.device(), ", ", self_.device(), ", ",
+              batch1.device(), " and ", batch2.device());
+  TORCH_CHECK(self_.dim() == 2,
+              "2D tensor expected, got ", self_.dim(), "D tensor for input");
+  int64_t batchnum = batch1.size(0);
+  int64_t m1d1 = batch1.size(1);
+  int64_t innerdim = batch1.size(2);
+  int64_t m2d2 = batch2.size(2);
+  TORCH_CHECK(batchnum == batch2.size(0),
+              "equal number of batches expected");
+  TORCH_CHECK(m1d1 == self_.size(0),
+              "first dimension of batch1  must match first dimension of input");
+  TORCH_CHECK(m2d2 == self_.size(1),
+              "second dimension of batch2 must match second dimension of input");
+  TORCH_CHECK(innerdim == batch2.size(1),
+              "second dimension of batch1 must match first dimension of batch2");
+
+  if (&out != &self) {
+    at::native::resize_as_(out, self_);
+    if (beta.to<double>() != 0.0) {
+      at::native::copy_(out, self_);
+    }
+  }
+
+  for (int64_t i=0; i<batchnum; i++) {
+    addmm_out_cuda(out, out, batch1[i], batch2[i], beta, alpha);
```

Comment:
I am not sure if I understand wrongly, but if we have `beta != 1`, will we end up having 
```
out = beta * out + alpha * (batch1[0] @ batch2[0])
out = beta * out + alpha * (batch1[1] @ batch2[1])
out = beta * out + alpha * (batch1[2] @ batch2[2])
```
where `beta` multiplied to the original `out` many times, instead of just once, which is different from the equation in https://pytorch.org/docs/stable/torch.html#torch.addbmm.

Reply:
One line below that, `beta` is set to `1` at the end of the first iteration. So we will get:

```
out = beta * out + alpha * (batch1[0] @ batch2[0])
out = 1 * out + alpha * (batch1[1] @ batch2[1])
out = 1 * out + alpha * (batch1[2] @ batch2[2])
```

Reply:
I see, thanks!